internal/util/ssa/apply.go

```diff
@@ -0,0 +1,60 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ssa
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+)
+
+// Run runs server side apply.
+// FIXME: sbueringer.
+func Run[T client.Object](ctx context.Context, c client.Client, requestTracker RequestTracker, original, modified T, fieldManager string) (T, error) {
```

Comment:
what about changing this signature so the caller can do something like
```
ssa.Patch(ctx, c, fieldManager, original, modified, WithCachingProxy(x))
```
Note: I like the idea of being explicit about when we want to use caching to avoid hitting the server, and I'm also assuming that when would like to use ssa.Patch for the calls that are not going about caching

Reply:
Have to check. I migrated everything to this func except the create cases.

So I assume you want to use this func for create as well? (create is usually just a very simple call)

And basically we have the effect that:
* for create this func is always called with modified = nil and without WithCachingProxy (otherwise we would panic)
* for update this func is always called with modified != nil and with WithCachingProxy